In [1]:
# Import needed libraries
import pandas as pd
import numpy as np
from plotnine import *

In [2]:
# Read in the dataset and show first few rows
md = pd.read_csv("/content/money_diaries.csv")
md.head()

Date Day of Week   City/Region State        Country    Income  \
0  01/24/2025      Friday    Pittsburgh    PA  United States  102750.0   
1  01/22/2025   Wednesday  Philadelphia    PA  United States  374000.0   
2  01/20/2025      Monday      New York    NY  United States  107141.0   
3  01/17/2025      Friday      New York    NY  United States  103000.0   
4  01/15/2025   Wednesday   Los Angeles    CA  United States  137000.0   

  Individual Income Joint       Industry  Age  ... Rents or Owns Total Assets  \
0            102750   Yes       Academia   27  ...         Rents    245000.00   
1            190000   Yes     Government   51  ...         Rents   2532405.00   
2            107141    No     Technology   25  ...         Rents    171392.44   
3            103000    No          Media   41  ...         Rents     90244.08   
4             67000   Yes  Environmental   27  ...         Rents     23350.00   

      Debt  Housing Costs per Month  Total Spent in Week  R29 Love Reactions  \
0      0.0                   2860.0               272.65                  30   
1  11542.0                   2625.0              1072.83                  98   
2      0.0                   2600.0               633.07                  50   
3      0.0                   2250.0              1062.89                  68   
4      0.0                   1650.0               242.14                 149   

   R29 Impressed Reactions  R29 Surprised Reactions  R29 Angry Reactions  \
0                        4                        1                   20   
1                        7                       13                   52   
2                       11                        9                   49   
3                       11                        9                   21   
4                        9                        2                    1   

   R29 Number of Comments  
0                      32  
1                      90  
2                      47  
3                      73  
4                      23  

[5 rows x 21 columns]

In [3]:
# Convert "Date" column to datetime format
md["Date"] = pd.to_datetime(md["Date"])
md.head()

Date Day of Week   City/Region State        Country    Income  \
0 2025-01-24      Friday    Pittsburgh    PA  United States  102750.0   
1 2025-01-22   Wednesday  Philadelphia    PA  United States  374000.0   
2 2025-01-20      Monday      New York    NY  United States  107141.0   
3 2025-01-17      Friday      New York    NY  United States  103000.0   
4 2025-01-15   Wednesday   Los Angeles    CA  United States  137000.0   

  Individual Income Joint       Industry  Age  ... Rents or Owns Total Assets  \
0            102750   Yes       Academia   27  ...         Rents    245000.00   
1            190000   Yes     Government   51  ...         Rents   2532405.00   
2            107141    No     Technology   25  ...         Rents    171392.44   
3            103000    No          Media   41  ...         Rents     90244.08   
4             67000   Yes  Environmental   27  ...         Rents     23350.00   

      Debt  Housing Costs per Month  Total Spent in Week  R29 Love Reactions  \
0      0.0                   2860.0               272.65                  30   
1  11542.0                   2625.0              1072.83                  98   
2      0.0                   2600.0               633.07                  50   
3      0.0                   2250.0              1062.89                  68   
4      0.0                   1650.0               242.14                 149   

   R29 Impressed Reactions  R29 Surprised Reactions  R29 Angry Reactions  \
0                        4                        1                   20   
1                        7                       13                   52   
2                       11                        9                   49   
3                       11                        9                   21   
4                        9                        2                    1   

   R29 Number of Comments  
0                      32  
1                      90  
2                      47  
3                      73  
4                      23  

[5 rows x 21 columns]

In [4]:
# Check data types of each column
md.dtypes

Date                       datetime64[ns]
Day of Week                        object
City/Region                        object
State                              object
Country                            object
Income                            float64
Individual Income                  object
Joint                              object
Industry                           object
Age                                 int64
Lives Alone                        object
Rents or Owns                      object
Total Assets                      float64
Debt                              float64
Housing Costs per Month           float64
Total Spent in Week               float64
R29 Love Reactions                  int64
R29 Impressed Reactions             int64
R29 Surprised Reactions             int64
R29 Angry Reactions                 int64
R29 Number of Comments              int64
dtype: object

In [5]:
# The "Individual Income" column is listed as an object, when it should be numeric. Let's see where the issue is

# md[md["Individual Income"].apply(lambda x: isinstance(x, object))] (This line of code revealed that all values in the column are objects)

# Now let's see all the unique values so we know which values caused the column to be rendered as strings
md["Individual Income"].unique()

array(['102750', '190000', '107141', '103000', '67000', '150000',
       '122000', '500000', '69319', '98000', '165000', '116820', '90000',
       '65238.45', '45000', '380000', '78000', '81000', '79500', '100000',
       '50000', '24000', '87500', '70000', '1600', '110000', '87000',
       '28000', '53501', '139000', '60000', '41000', '0', '80000',
       '120000', '29791.77', '275000', '130000', '215000', '56000',
       '44600', '35000', '127000', '255000', '57398', '175000', '228000',
       '30000', '85000', '135000', '126000', '140000', '114000', '95000',
       '160000', '115000', '100296', '48000', '97500', '48946', '82000',
       '116000', '108280', '36000', 'Unknown', '82300', '109000',
       '204000', '67525.05', '134000', '171000', '144200', '240000',
       '261000', '131858', '192000', '137000', '116500', '205000',
       '125000', '118000', '310000', '59200', '58000', '98300', '179000',
       '73500', '43000', '108150', '65000', '55834.01', '65692.14',
       '225000'

In [6]:
# The "Unknown" value is what's causing the issue. Let's remove all row(s) with "Unknown in that column", then convert the column to float
import warnings
warnings.filterwarnings("ignore", category=pd.errors.SettingWithCopyWarning)

md = md[md["Individual Income"] != "Unknown"]

md["Individual Income"] = md["Individual Income"].astype(float)

### Let's see all the unique values for the categorical columns

In [7]:
# md["Day of Week"].value_counts() # Only one wasn't posted on a Monday, Wednesday, or Friday, so I think I will delete this row

md = md[md["Day of Week"] != "Tuesday"]
md["Day of Week"].value_counts() # just to double-check that it's fixed

Day of Week
Friday       71
Monday       70
Wednesday    67
Name: count, dtype: int64

In [8]:
md["City/Region"].unique() # some regions overlap (such as "Bay Area" and "San Francisco"), so this may present an issue in the model

array(['Pittsburgh', 'Philadelphia', 'New York', 'Los Angeles',
       'Detroit suburbs', 'Bay Area', 'Central Minnesota', 'Mexico City',
       'Boston', 'Toronto', 'Kansas City Metro', 'Raleigh', 'Montana',
       'Providence', 'Indianapolis', 'Tampa', 'Seoul',
       'Northern California', 'Chicago', 'Michigan', 'Hampton Roads',
       'San Diego', 'Central Florida', 'San Francisco',
       'Central Pennsylvania', 'Atlantic Canada', 'Cincinnati',
       'Latin America', 'Arkansas', 'DMV', 'Virginia',
       'Greater Cincinnati', 'Oregon', 'Fort Garland', 'Nashville',
       'NYC Suburbs', 'Orange County', 'Brooklyn', 'New Jersey',
       'Richmond', 'Kansas City', 'El Paso', 'Baltimore', 'Boise',
       'Wilmington', 'Milwaukee', 'Tri-Cities', 'Denver',
       'Washington, D.C.', "O'ahu", 'Montreal', 'Oakland', 'Queens',
       'Miami', 'Kansas City suburbs', 'Austin', 'South Carolina',
       'Dallas', 'Phoenix', 'Burtonsville', 'Grand Rapids', 'Portland',
       'Huntsville', 'Van

In [9]:
# Let's fix this
md["City/Region"] = md["City/Region"].replace({"Bay Area": "San Francisco Bay Area", "San Francisco": "San Francisco Bay Area",
                                               "Oakland": "San Francisco Bay Area", "NYC Suburbs": "New York", "Brooklyn": "New York",
                                               "Queens": "New York", "Kansas City Metro": "Kansas City", "Kansas City suburbs":
                                               "Kansas City", "Detroit suburbs": "Detroit", "Greater Cincinnati": "Cincinnati", "Chicago suburb":
                                               "Chicago", "DMV": "Washington, D.C."})

# thankfully all of the Kansas City values refer to Kansas City, KS; there are no Kansas City, MO values in the dataset, which prevents confusion in
# the model

md["City/Region"].unique() # check

array(['Pittsburgh', 'Philadelphia', 'New York', 'Los Angeles', 'Detroit',
       'San Francisco Bay Area', 'Central Minnesota', 'Mexico City',
       'Boston', 'Toronto', 'Kansas City', 'Raleigh', 'Montana',
       'Providence', 'Indianapolis', 'Tampa', 'Seoul',
       'Northern California', 'Chicago', 'Michigan', 'Hampton Roads',
       'San Diego', 'Central Florida', 'Central Pennsylvania',
       'Atlantic Canada', 'Cincinnati', 'Latin America', 'Arkansas',
       'Washington, D.C.', 'Virginia', 'Oregon', 'Fort Garland',
       'Nashville', 'Orange County', 'New Jersey', 'Richmond', 'El Paso',
       'Baltimore', 'Boise', 'Wilmington', 'Milwaukee', 'Tri-Cities',
       'Denver', "O'ahu", 'Montreal', 'Miami', 'Austin', 'South Carolina',
       'Dallas', 'Phoenix', 'Burtonsville', 'Grand Rapids', 'Portland',
       'Huntsville', 'Vancouver Island', 'Arlington', 'Des Moines',
       'Houston', 'Madison', 'California', 'Illinois', 'Ottawa', 'Salem',
       'Minneapolis', 'Bellingham', 

In [10]:
md["State"].unique()

array(['PA', 'NY', 'CA', 'MI', 'MN', nan, 'MA', 'KS', 'NC', 'MT', 'RI',
       'IN', 'FL', 'IL', 'VA', 'OH', 'AR', 'Unknown', 'OR', 'CO', 'TN',
       'NJ', 'TX', 'MD', 'ID', 'DE', 'WI', 'WA', 'Washington, D.C.', 'HI',
       'SC', 'AZ', 'AL', 'IA', 'VT', 'GA', 'KY', 'UT', 'NM', 'ME'],
      dtype=object)

In [11]:
# Let's replace all nan values in the "State" column with a different value
md["State"] = md["State"].fillna("Not U.S.")
md["State"].unique() # check

array(['PA', 'NY', 'CA', 'MI', 'MN', 'Not U.S.', 'MA', 'KS', 'NC', 'MT',
       'RI', 'IN', 'FL', 'IL', 'VA', 'OH', 'AR', 'Unknown', 'OR', 'CO',
       'TN', 'NJ', 'TX', 'MD', 'ID', 'DE', 'WI', 'WA', 'Washington, D.C.',
       'HI', 'SC', 'AZ', 'AL', 'IA', 'VT', 'GA', 'KY', 'UT', 'NM', 'ME'],
      dtype=object)

In [12]:
md["Country"].value_counts()

Country
United States     191
Canada             11
Mexico              1
South Korea         1
Unknown             1
Venezuela           1
United Kingdom      1
Luxembourg          1
Name: count, dtype: int64

In [13]:
# It's probably simpler to categorize the countries as US, Canada, or "not US or Canada"
md["Country"] = md["Country"].apply(lambda x: x if x in ["United States", "Canada"] else "Not United States or Canada")

md["Country"].value_counts() # checking to make sure it worked as intended

Country
United States                  191
Canada                          11
Not United States or Canada      6
Name: count, dtype: int64

In [14]:
# I also checked the unique values of the other string columns to ensure there were no misspelled values or other errors

# md["Joint"].unique()
# md["Lives Alone"].unique()
# md["Rents or Owns"].unique()

In [15]:
# Let's clean up the "Industry" column a bit
  # There can be a lot of confusion here, especially since one diarist might list their industry as "tech" while another might say "technology",
  # or they might list two industries at once

md["Industry"].unique()

array(['Academia', 'Government', 'Technology', 'Media', 'Environmental',
       'Education, Mental Health', 'Finance', 'Education', 'Nonprofit',
       'Media, Publishing', 'Retail', 'Accounting', 'Higher education',
       'Law', 'Software', 'Humanities studies', 'Food industry',
       'Pharma advertising', 'Healthcare',
       'Local government, library services', 'Climate tech',
       'Nonprofit, heritage', 'Health',
       'Specialty retail, customer service/security, entertainment',
       'Aviation', 'General appraising, cultural history, and crafts',
       'Manufacturing', 'Pharmaceuticals', 'Retail trade',
       'Medical, nonprofit', 'Biotech', 'Health insurance', 'Automotive',
       nan, 'Transportation and logistics', 'Tech', 'Loan servicing',
       'Transportation', 'IT', 'Consulting', 'Emergency management',
       'Advertising', 'Corporate healthcare', 'Energy', 'Real estate',
       'Marketing', 'Utilities', 'Media and entertainment',
       'Tech startup', 'Private

In [16]:
# First, let's replace the "nan" values with "Unemployed" since unemployment/no income is the reason for the nan values
md["Industry"] = md["Industry"].fillna("Unemployed")
md["Industry"].unique() # check

array(['Academia', 'Government', 'Technology', 'Media', 'Environmental',
       'Education, Mental Health', 'Finance', 'Education', 'Nonprofit',
       'Media, Publishing', 'Retail', 'Accounting', 'Higher education',
       'Law', 'Software', 'Humanities studies', 'Food industry',
       'Pharma advertising', 'Healthcare',
       'Local government, library services', 'Climate tech',
       'Nonprofit, heritage', 'Health',
       'Specialty retail, customer service/security, entertainment',
       'Aviation', 'General appraising, cultural history, and crafts',
       'Manufacturing', 'Pharmaceuticals', 'Retail trade',
       'Medical, nonprofit', 'Biotech', 'Health insurance', 'Automotive',
       'Unemployed', 'Transportation and logistics', 'Tech',
       'Loan servicing', 'Transportation', 'IT', 'Consulting',
       'Emergency management', 'Advertising', 'Corporate healthcare',
       'Energy', 'Real estate', 'Marketing', 'Utilities',
       'Media and entertainment', 'Tech startup',

In [17]:
# Now let's categorize industries to get a clearer idea of trends in R29 comments activity
md["Industry"] = md["Industry"].map({
    "Academia": "Education", "Education": "Education", "Higher education": "Education", "Humanities studies": "Education", "University": "Education",

    "Government": "Government", "Research and government contract": "Government", "Federal government": "Government", "Public sector": "Government",
    "Politics": "Government", "Public service": "Government",

    "Technology": "Tech", "Software": "Tech", "Climate tech": "Tech", "Tech": "Tech", "IT": "Tech", "Tech startup": "Tech", "Fintech": "Tech",
    "Cybersecurity": "Tech",

    "Media": "Communications", "Media, Publishing": "Communications", "Pharma advertising": "Communications", "Advertising": "Communications",
    "Marketing": "Communications", "Media and entertainment": "Communications", "Entertainment": "Communications",
    "Book publishing": "Communications", "TV/media": "Communications", "News media": "Communications", "Telecommunications": "Communications",
    "Marketing/entertainment": "Communications", "Communications": "Communications", "Public relations": "Communications",
    "Film and TV": "Communications", "News": "Communications", "Brand strategy": "Communications", "E-commerce": "Communications",

    "Environmental": "Science", "Biotech": "Science", "Emergency management": "Science", "Life sciences": "Science",

    "Education, Mental Health": "Various", "Specialty retail, customer service/security, entertainment": "Various",
    "Education and healthcare": "Various", "Fitness/events": "Various", "Higher education/retail": "Various", "Medical/Tech": "Various",
    "Various": "Various", "Commercial aviation and automotive marketing": "Various",

    "Finance": "Finance", "Accounting": "Finance", "Health insurance": "Finance", "Loan servicing": "Finance", "Private equity": "Finance",
    "Investment management": "Finance", "Financial services": "Finance", "Insurance": "Finance",

    "Nonprofit": "Nonprofit", "Nonprofit, heritage": "Nonprofit", "Medical, nonprofit": "Nonprofit", "Non-profit": "Nonprofit",

    "Retail": "Retail", "Retail trade": "Retail",

    "Law": "Law", "Legal": "Law",

    "Food industry": "Food industry",

    "Healthcare": "Healthcare", "Health": "Healthcare", "Corporate healthcare": "Healthcare", "Healthcare, intensive care": "Healthcare",
    "Healthcare tech": "Healthcare", "Medicine": "Healthcare",

    "Local government, library services": "Arts", "General appraising, cultural history, and crafts": "Arts", "Marketing and design": "Arts",
    "Museums/libraries": "Arts", "Arts": "Arts", "Art": "Arts",

    "Aviation": "Transportation", "Automotive": "Transportation", "Transportation and logistics": "Transportation",
    "Transportation": "Transportation",

    "Pharmaceuticals": "Pharmaceuticals",

    "Unemployed": "Unemployed",

    "Consulting": "Consulting", "Tech consulting": "Consulting", "Environmental consulting": "Consulting",

    "Real estate": "Real estate", "Commercial real estate development": "Real estate",

    "Oil and gas": "Oil and gas", "Oil and gas drilling": "Oil and gas",

    "Consumer packaged goods": "Consumer packaged goods",

    "Manufacturing": "Other", "Energy": "Other", "Utilities": "Other", "Labor": "Other", "Regulatory": "Other", "Architecture": "Other",
    "Fashion": "Other", "Sales": "Other", "International development": "Other", "Construction": "Other"
})

md["Industry"].unique() # check

array(['Education', 'Government', 'Tech', 'Communications', 'Science',
       'Various', 'Finance', 'Nonprofit', 'Retail', 'Law',
       'Food industry', 'Healthcare', 'Arts', 'Transportation', 'Other',
       'Pharmaceuticals', 'Unemployed', 'Consulting', 'Real estate',
       'Oil and gas', 'Consumer packaged goods'], dtype=object)

In [18]:
# Let's create a net worth column too
md["Net Worth"] = md["Total Assets"] - md["Debt"]

md[["Net Worth", "Total Assets", "Debt"]].head()

Net Worth  Total Assets     Debt
0   245000.00     245000.00      0.0
1  2520863.00    2532405.00  11542.0
2   171392.44     171392.44      0.0
3    90244.08      90244.08      0.0
4    23350.00      23350.00      0.0

In [19]:
# Now that the dataset is cleaned, let's save it as a new csv that will be used for Tableau visualizations
md.to_csv("money_diaries_clean.csv", index=False)

Engagement (measured in the number of comments on each diary) has been on an overall decline since September 2023. The months with the lowest average engagement in any year are May, July, and August. Day of week doesn't significantly impact engagement, though it does decline a bit throughout the average week (Monday diaries get an average of 74 comments while Friday diaries get an average of 64 comments).

Let's use predictive modeling in scikit-learn to see which factors drive engagement the most, to see which diaries Refinery29 may want to prioritize when deciding which of their submissions to publish

### Let's try linear regression to start, with R29 Number of Comments being the target variable

In [20]:
md.head()

Date Day of Week   City/Region State        Country    Income  \
0 2025-01-24      Friday    Pittsburgh    PA  United States  102750.0   
1 2025-01-22   Wednesday  Philadelphia    PA  United States  374000.0   
2 2025-01-20      Monday      New York    NY  United States  107141.0   
3 2025-01-17      Friday      New York    NY  United States  103000.0   
4 2025-01-15   Wednesday   Los Angeles    CA  United States  137000.0   

   Individual Income Joint        Industry  Age  ... Total Assets     Debt  \
0           102750.0   Yes       Education   27  ...    245000.00      0.0   
1           190000.0   Yes      Government   51  ...   2532405.00  11542.0   
2           107141.0    No            Tech   25  ...    171392.44      0.0   
3           103000.0    No  Communications   41  ...     90244.08      0.0   
4            67000.0   Yes         Science   27  ...     23350.00      0.0   

   Housing Costs per Month  Total Spent in Week  R29 Love Reactions  \
0                   2860.0               272.65                  30   
1                   2625.0              1072.83                  98   
2                   2600.0               633.07                  50   
3                   2250.0              1062.89                  68   
4                   1650.0               242.14                 149   

   R29 Impressed Reactions  R29 Surprised Reactions  R29 Angry Reactions  \
0                        4                        1                   20   
1                        7                       13                   52   
2                       11                        9                   49   
3                       11                        9                   21   
4                        9                        2                    1   

   R29 Number of Comments   Net Worth  
0                      32   245000.00  
1                      90  2520863.00  
2                      47   171392.44  
3                      73    90244.08  
4                      23    23350.00  

[5 rows x 22 columns]

In [43]:
from sklearn.pipeline import Pipeline
from sklearn.compose import make_column_selector, ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, PolynomialFeatures
from sklearn.linear_model import LinearRegression

X = md.drop(["Date", "R29 Number of Comments"], axis = 1)
y = md["R29 Number of Comments"]

ct = ColumnTransformer(
  [
    ("dummify",
    OneHotEncoder(sparse_output = False, handle_unknown="ignore"),
    make_column_selector(dtype_include=object)),
    ("standardize",
    StandardScaler(),
    make_column_selector(dtype_include=np.number))
  ],
  remainder = "passthrough"
)

lr_pipeline = Pipeline(
  [("preprocessing", ct),
  ("linear_regression", LinearRegression())]
)

In [47]:
# Fit linear regression pipeline on full dataset and interpret some of the important coefficients
lr_pipeline_fitted = lr_pipeline.fit(X, y)
lr_coefs = lr_pipeline_fitted.named_steps["linear_regression"].coef_

feature_names = ct.get_feature_names_out()
pd.set_option("display.max_rows", None)
pd.set_option("display.float_format", "{:.6f}".format) # makes sure output is not in scientific notation
pd.Series(abs(lr_coefs), index=feature_names).sort_values(ascending=False) # I'm viewing the coefficients by absolute numbers first,
                                                                            # then once I narrow down the most important features, I will
                                                                            # set them back to positive and negative

dummify__Industry_Oil and gas                  145.035145
dummify__City/Region_Rural Colorado            140.084007
dummify__City/Region_Bellingham                134.379514
dummify__City/Region_Indiana                    87.580417
dummify__City/Region_Westchester                76.386691
dummify__City/Region_Denver                     69.537951
dummify__City/Region_Central Minnesota          64.741703
dummify__City/Region_Detroit                    58.651119
dummify__Industry_Consulting                    51.251630
dummify__State_CO                               47.146906
dummify__City/Region_Indianapolis               46.295443
dummify__City/Region_Hampton Roads              43.078872
dummify__City/Region_Virginia                   42.162257
dummify__State_IN                               41.284974
dummify__City/Region_Montreal                   40.946581
dummify__Industry_Unemployed                    39.620581
dummify__City/Region_Illinois                   39.185736
dummify__City/Region_New York                   38.309981
dummify__Industry_Healthcare                    37.254530
dummify__City/Region_Northern Virginia          36.186234
dummify__City/Region_Luxembourg                 33.877067
dummify__City/Region_Minneapolis                32.562792
dummify__City/Region_Houston                    31.429478
dummify__City/Region_Richmond                   30.420075
dummify__Industry_Transportation                30.090211
dummify__Industry_Government                    29.539944
dummify__State_MI                               29.417989
dummify__Industry_Consumer packaged goods       28.650997
standardize__R29 Angry Reactions                28.601489
dummify__City/Region_Latin America              28.504659
dummify__City/Region_Central Florida            28.500074
dummify__City/Region_Palm Springs               27.557004
dummify__City/Region_Vancouver Island           27.536634
dummify__City/Region_Atlanta                    26.033876
dummify__State_GA                               26.033876
dummify__Industry_Real estate                   25.571735
dummify__City/Region_Phoenix                    25.242875
dummify__City/Region_Salem                      24.916129
dummify__City/Region_Vancouver                  24.456724
dummify__City/Region_Fort Garland               23.399150
dummify__City/Region_El Paso                    23.373065
dummify__Industry_Other                         23.169660
dummify__City/Region_Orange County              22.817625
dummify__City/Region_Miami                      22.749626
dummify__State_TX                               21.923620
dummify__State_IL                               21.858037
dummify__State_MN                               21.735474
dummify__City/Region_Louisville                 21.413494
standardize__Income                             20.484945
dummify__City/Region_Portland                   19.998185
dummify__City/Region_Wilmington                 19.746999
dummify__State_DE                               19.746999
dummify__City/Region_Charlotte                  19.737834
dummify__City/Region_Michigan                   19.711893
dummify__State_NY                               19.682842
dummify__City/Region_Cincinnati                 19.674610
dummify__Industry_Retail                        19.376347
dummify__State_FL                               19.221817
dummify__Industry_Food industry                 19.003202
dummify__City/Region_Dallas                     18.824027
dummify__City/Region_California                 18.556425
dummify__City/Region_Pittsburgh                 18.457112
dummify__City/Region_Northern Kentucky          18.422073
dummify__City/Region_Vermont                    18.271128
dummify__State_VT                               18.271128
dummify__State_AR                               18.058876
dummify__City/Region_Arkansas                   18.058876
dummify__City/Region_Northern California        17.521571
dummify__City/Region_Chicago                    17.32769

In [35]:
# Use cross-validation to estimate R-squared
from sklearn.model_selection import cross_val_score
scores = cross_val_score(lr_pipeline, X, y, cv=5, scoring="r2")
scores.mean()

np.float64(-0.645121476308008)

In [49]:
# the top 5 variables seem to be Industry, City/Region, State, R29 Angry Reactions, and Income. Let's try a model with just those
X = md[["Industry", "City/Region", "State", "R29 Angry Reactions", "Income"]]
y = md["R29 Number of Comments"]

ct = ColumnTransformer(
  [
    ("dummify",
    OneHotEncoder(sparse_output = False, handle_unknown="ignore"),
    make_column_selector(dtype_include=object)),
    ("standardize",
    StandardScaler(),
    make_column_selector(dtype_include=np.number))
  ],
  remainder = "passthrough"
)

lr_pipeline = Pipeline(
  [("preprocessing", ct),
  ("linear_regression", LinearRegression())]
)

scores = cross_val_score(lr_pipeline, X, y, cv=5, scoring="r2")
scores.mean()

np.float64(-0.5029282697447306)

In [50]:
# Much better than the one with all the variables! Now let's try the top 10, which include Joint, R29 Impressed Reactions, Age, Day of Week &
                                                                                                                          # Individual Income
X = md[["Industry", "City/Region", "State", "R29 Angry Reactions", "Income", "Joint", "R29 Impressed Reactions", "Age", "Day of Week",
        "Individual Income"]]
y = md["R29 Number of Comments"]

ct = ColumnTransformer(
  [
    ("dummify",
    OneHotEncoder(sparse_output = False, handle_unknown="ignore"),
    make_column_selector(dtype_include=object)),
    ("standardize",
    StandardScaler(),
    make_column_selector(dtype_include=np.number))
  ],
  remainder = "passthrough"
)

lr_pipeline = Pipeline(
  [("preprocessing", ct),
  ("linear_regression", LinearRegression())]
)

scores = cross_val_score(lr_pipeline, X, y, cv=5, scoring="r2")
scores.mean()

np.float64(-0.3800777812901016)

In [51]:
# Even better! Now let's do the top 15
X = md[["Industry", "City/Region", "State", "R29 Angry Reactions", "Income", "Joint", "R29 Impressed Reactions", "Age", "Day of Week",
        "Individual Income", "Lives Alone", "Debt", "Housing Costs per Month", "R29 Surprised Reactions", "Total Spent in Week"]]
y = md["R29 Number of Comments"]

ct = ColumnTransformer(
  [
    ("dummify",
    OneHotEncoder(sparse_output = False, handle_unknown="ignore"),
    make_column_selector(dtype_include=object)),
    ("standardize",
    StandardScaler(),
    make_column_selector(dtype_include=np.number))
  ],
  remainder = "passthrough"
)

lr_pipeline = Pipeline(
  [("preprocessing", ct),
  ("linear_regression", LinearRegression())]
)

scores = cross_val_score(lr_pipeline, X, y, cv=5, scoring="r2")
scores.mean()

np.float64(-0.6261112655448613)

That one's almost as bad as the one with all the variables! I experimented by removing each of the last five one by one to see which ones were causing the score to go down.

The best feature set I could come up with was the top 10 plus Debt, which led to roughly an R-squared of -0.3544.

In [74]:
X = md[["Industry", "City/Region", "State", "R29 Angry Reactions", "Income", "Joint", "R29 Impressed Reactions", "Age", "Day of Week",
        "Individual Income", "Debt"]]
y = md["R29 Number of Comments"]

ct = ColumnTransformer(
  [
    ("dummify",
    OneHotEncoder(sparse_output = False, handle_unknown="ignore"),
    make_column_selector(dtype_include=object)),
    ("standardize",
    StandardScaler(),
    make_column_selector(dtype_include=np.number))
  ],
  remainder = "passthrough"
)

lr_pipeline = Pipeline(
  [("preprocessing", ct),
  ("linear_regression", LinearRegression())]
)

scores = cross_val_score(lr_pipeline, X, y, cv=5, scoring="r2")
scores.mean()

np.float64(-0.3544347253475079)

### Let's try this same feature set but with Ridge regression

In [75]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV
ridge_pipeline = Pipeline(
  [("preprocessing", ct), # the same preprocessing ColumnTransformer as earlier works
  ("ridge", Ridge())]
)

# Use cross-validation to tune λ
param_grid = {
    "ridge__alpha": [0.001, 0.01, 0.1, 1, 10]
}

gscv = GridSearchCV(ridge_pipeline, param_grid, cv = 5, scoring="r2")
gscv_fitted = gscv.fit(X, y)
gscv_fitted.cv_results_["mean_test_score"]

array([-0.35118544, -0.32483719, -0.20167692, -0.08126596,  0.03143012])

A Ridge regression model with the best feature set and λ=10 would have an R-squared of 0.0314, a significant improvement from the linear regression model's -0.3544!

### Now let's try LASSO

In [78]:
from sklearn.linear_model import Lasso
import warnings
from sklearn.exceptions import ConvergenceWarning
warnings.filterwarnings("ignore", category=ConvergenceWarning) # ConvergenceWarnings are creating noisy output


lasso_pipeline = Pipeline(
  [("preprocessing", ct),
  ("lasso", Lasso())]
)

param_grid = {
    "lasso__alpha": [0.001, 0.01, 0.1, 1, 10]
}

gscv = GridSearchCV(lasso_pipeline, param_grid, cv = 5, scoring="r2")
gscv_fitted = gscv.fit(X, y)
gscv_fitted.cv_results_["mean_test_score"]

array([-0.24965035, -0.18428461,  0.03357745,  0.05982828, -0.06258748])

A LASSO regression model with the best feature set and λ=1 would have an R-squared of 0.0598, the best so far!

### Finally, let's see if Elastic Net will give us the best of both worlds

In [80]:
from sklearn.linear_model import ElasticNet

elastic_net_pipeline = Pipeline(
  [("preprocessing", ct),
  ("elastic_net", ElasticNet())])

param_grid = {
    "elastic_net__alpha": [0.001, 0.01, 0.1, 1, 10],
    "elastic_net__l1_ratio": [0.1, 0.3, 0.5, 0.7, 0.9, 1.0]
}

gscv = GridSearchCV(elastic_net_pipeline, param_grid, cv=5, scoring="r2")
gscv_fitted = gscv.fit(X, y)
print("Best params:", gscv_fitted.best_params_)
print("Best score:", gscv_fitted.best_score_)

Best params: {'elastic_net__alpha': 1, 'elastic_net__l1_ratio': 0.9}
Best score: 0.06980924244558806
